Patch originally from KDE commit d1a1c80370e4132c26fa636308822fdb3aa1b2e6, modified to work with current Amarok.
diff -ruN /root/amarok-2.5.0/ChangeLog amarok-2.5.0/ChangeLog
--- a/amarok-2.5.0/ChangeLog	2011-12-15 02:36:48.000000000 -0500
+++ b/amarok-2.5.0/ChangeLog	2012-06-21 21:05:29.871236927 -0400
@@ -8,6 +8,7 @@
       Playlist's layout items.
 
   BUGFIXES:
+    * Fix crash on start when Qt is build with debugging assertions. (BR 285720)
     * Do not crash when iPod doesn't have a master playlist. (BR 288936)
     * Fix Last.fm not finding KDE Wallet. (BR 286741)
     * Put a border around bottom toolbars, visually separating them from
diff -ruN /root/amarok-2.5.0/src/browsers/playlistbrowser/QtGroupingProxy.cpp amarok-2.5.0/src/browsers/playlistbrowser/QtGroupingProxy.cpp
--- a/amarok-2.5.0/src/browsers/playlistbrowser/QtGroupingProxy.cpp	2011-12-15 02:36:48.000000000 -0500
+++ b/amarok-2.5.0/src/browsers/playlistbrowser/QtGroupingProxy.cpp	2012-06-21 21:05:09.810237474 -0400
@@ -223,23 +223,15 @@
         {
             int &rowValue = groupList[insertedProxyRow-1];
             if( idx.row() <= rowValue )
-            {
                 //increment the rows that come after the new row since they moved one place up.
                 rowValue++;
-            }
             else
-            {
                 break;
-            }
         }
 
         if( updatedGroups.contains( i.key() ) )
-        {
-            //the row needs to be added to this group
-            beginInsertRows( index( i.key() ), insertedProxyRow, insertedProxyRow );
-            groupList.insert( insertedProxyRow, idx.row() );
-            endInsertRows();
-        }
+            // we're inside beginInsertRows() or beginInsertRows(), don't re-enter it.
+	     groupList.insert( insertedProxyRow, idx.row() );
     }
 
     return updatedGroups;
diff -ruN /root/amarok-2.5.0/src/browsers/playlistbrowser/QtGroupingProxy.h amarok-2.5.0/src/browsers/playlistbrowser/QtGroupingProxy.h
--- a/amarok-2.5.0/src/browsers/playlistbrowser/QtGroupingProxy.h	2011-12-15 02:36:48.000000000 -0500
+++ b/amarok-2.5.0/src/browsers/playlistbrowser/QtGroupingProxy.h	2012-06-21 21:04:42.778238198 -0400
@@ -38,8 +38,8 @@
         void setGroupedColumn( int groupedColumn );
 
         /* QAbstractProxyModel methods */
-        virtual QModelIndex index( int, int c = 0,
-                                   const QModelIndex& parent = QModelIndex() ) const;
+	virtual QModelIndex index( int row, int column = 0,
+				    const QModelIndex& parent = QModelIndex() ) const;
         virtual Qt::ItemFlags flags( const QModelIndex &idx ) const;
         virtual QModelIndex parent( const QModelIndex &idx ) const;
         virtual int rowCount( const QModelIndex &idx = QModelIndex() ) const;
